### DISCUSSION TITLE
Plugin + TypeScript: How to type new state properties based on new option?

### DISCUSSION THREAD

**Discussion by ByScripts** [vuejs/pinia#1411] [Help and Questions]
Hi,

I'm trying to get custom prop typing based on custom option value.

Let's take this simple plugin:

```ts
function myPlugin({ options, store }: PiniaPluginContext) {
  if (options.addCustomProp) {
    store.customProp = 'my custom prop;
  }
}
```

It adds a custom prop to the state only if the `addCustomProp` is set to true.

For now, I added these declarations:

```ts
declare module 'pinia' {
  export interface PiniaCustomProperties {
    customProp: string
  }

  export interface DefineStoreOptionsBase {
    addCustomProp?: boolean
  }
}
```

It works, BUT the `customProp` state property is now proposed for ALL stores, even if the `addCustomProp` option was not passed (or passed as `false`)

Is there a way to add `customProp` typing to `PiniaCustomStateProperties` based on the `addCustomProp` option value?

Thank you.

**Comment by ilyarefraction** [comment]
@ByScripts Check [this discussion](https://github.com/vuejs/pinia/discussions/2027). The only way is explicit check against `store.$id`
